src/v/cloud_topics/level_zero/stm/ctp_stm.cc

```diff
@@ -11,84 +11,141 @@
+}
 
-    // Note: do_apply will be called multiple times with the same batch if it
-    // throws. The method must be idempotent. Because of this we use the batch
-    // base offset as dl_version rather than record offset. This also means that
-    // in the same batch we cannot add an overlay and then remove it. Other
-    // caveats may exist as well.
-    //
-    // The version can't go backwards but in case of a partial apply and a retry
-    // it could.
-    auto new_dl_version = dl_version(batch.base_offset());
-
-    if (!_state.get_offsets().can_apply(new_dl_version)) {
-        vlog(
-          _log.warn,
-          "Record batch at offset {} is applied out of order",
-          new_dl_version);
+ss::future<> ctp_stm::do_apply(const model::record_batch& batch) {
+    if (
+      batch.header().type != model::record_batch_type::dl_placeholder
+      && batch.header().type != model::record_batch_type::ctp_stm_command) {
         co_return;
     }
+    vlog(_log.debug, "Applying record batch: {}", batch.header());
 
-    batch.for_each_record([new_dl_version, this](model::record&& r) {
-        auto key = serde::from_iobuf<ctp_stm_key>(r.release_key());
-        switch (key) {
-        case ctp_stm_key::start_snapshot: {
-            std::ignore = serde::from_iobuf<start_snapshot_cmd>(
-              r.release_value());
-            // Noexcept
-            _state.start_snapshot(new_dl_version);
-            break;
-        }
-        case ctp_stm_key::remove_snapshots_before_version:
-            auto cmd = serde::from_iobuf<remove_snapshots_before_version_cmd>(
-              r.release_value());
-            try {
-                _state.remove_snapshots_before(cmd.last_version_to_keep);
-            } catch (const std::runtime_error& e) {
-                // We don't have any other option but to ignore the error.
-                // The STM behaves deterministically so retrying will result
-                // in the same exception. The exception can only be caused
-                // by the incorrect command in the log (invalid version that
-                // can't be found).
-                vlog(
-                  _log.error,
-                  "'remove_snapshots_before_version command at @{} can't be "
-                  "applied because of error: {}",
-                  new_dl_version,
-                  e);
+    if (batch.header().type == model::record_batch_type::dl_placeholder) {
+        // Cherry-pick the placeholder from the record batch
+        vassert(
+          batch.record_count() > 0,
+          "Record batch must have at least one record");
+        iobuf value;
+        batch.for_each_record([&value](model::record&& r) {
+            value = std::move(r).release_value();
+            return ss::stop_iteration::yes;
+        });
+
+        auto placeholder = serde::from_iobuf<dl_placeholder>(std::move(value));
+        auto id = placeholder.id;
+        _state.advance_epoch(id.epoch);
```

Comment:
IIRC the plan was to use locking and term replication to guarantee the ordering of epoch values in the log. So here I would have expected to see only verification of that property, instead of the log replay seeming to drive the state change.